cspyce/spice_cell.py

```diff
@@ -0,0 +1,170 @@
+import numpy as np
+
+
+SPICE_CELL_HEADER_DESCRIPTOR = np.dtype([
+    # Many of these names clash with built-in properties of an array, so we've suffixed
+    # all of them with an underscore
+    ("dtype_", np.int32),
+    ("length_", np.int32),
+    ("size_", np.int32),
+    ("card_", np.int32),
+    ("isSet_", np.int32),
+    ("adjust_", np.int32),
+    ("init_", np.int32),
+    ("base_", np.int64),
+    ("data_", np.int64),
+], align=True)
+
+SPICE_CELL_INT = 2
+SPICE_CELL_DOUBLE = 1
+SPICE_CELL_CHAR = 0
+
+class SpiceCell:
+    CONTROL_SIZE = 6
+
+    @staticmethod
+    def create_spice_cell(my_type, size=1000, length=30):
+        length = length if my_type == SPICE_CELL_CHAR else 0
+        return SpiceCell(typeno=my_type, size=size, length=length)
+
+    @staticmethod
+    def as_spice_cell(my_type, record):
+        if isinstance(record, SpiceCell) and record._header.dtype_ == my_type:
+            return record
+        return SpiceCell(data=record, typeno=my_type)
+
+    def __init__(self, data=None, typeno=None, size: int = 0, length: int = 0):
+        if data is None:
+            if typeno is None or size == 0 or (type == SPICE_CELL_DOUBLE and length == 0):
+                raise ValueError("You must specify either an array, "
+                                 "or else type, size, and length (for characters)")
+
+        if typeno is None:
+            data = np.asarray(data)
+            if data.dtype.kind == 'f':
+                typeno = SPICE_CELL_DOUBLE
+            elif data.dtype.kind in 'bis':
+                typeno = SPICE_CELL_INT
+            elif data.dtype.kind in 'SU':
+                typeno = SPICE_CELL_CHAR
+                data = np.asarray(data, dtype='S')
+                length = max(2, length, max(len(x) for x in data.ravel()) + 1)
+            else:
+                raise ValueError("array has unknown type")
+
+        if typeno == SPICE_CELL_INT:
+            array_descriptor = np.dtype(np.int32)
+        elif typeno == SPICE_CELL_DOUBLE:
+            array_descriptor = np.dtype(np.double)
+        elif typeno == SPICE_CELL_CHAR:
+            array_descriptor = np.dtype(("S", length))
+            length = max(2, length)
+        else:
+            raise ValueError(f"Bad type {typeno} passed to SpiceCell init")
+
+        if data is not None:
+            data = np.asarray(data, dtype=array_descriptor)
+            size = max(size, len(data) + 6)  # add a little bit of spare room
+
+        self._header = np.rec.array(None, SPICE_CELL_HEADER_DESCRIPTOR, 1)[0]
+        self._descriptor = array_descriptor
+
+        self._header.dtype_ = typeno
```

Comment:
Why do these have the underscores at the end?

Reply:
See the comment on line 5.  Many of these names conflict with numpy internal names for arrays.  It was easier just to put a _ on all of them rather than try to keep track which ones needed it and which didn't.

Reply:
Why does this matter? SpiceCell isn't a subclass of `ndarray`. But these are private items anyway, right, since they're part of `_header`? So why not put the underscore at the beginning indicating they are private?